cpp/src_prims/selection/knn.h

```diff
@@ -104,93 +170,451 @@ void merge_tables(long n, long k, long nshard, float *distances, long *labels,
+          Distance::DistanceType DistanceType = Distance::EucUnexpandedL2>
+void brute_force_knn(std::vector<float *> &input, std::vector<int> &sizes,
+                     IntType D, float *search_items, IntType n, int64_t *res_I,
+                     float *res_D, IntType k,
+                     std::shared_ptr<deviceAllocator> allocator,
+                     cudaStream_t userStream,
+                     cudaStream_t *internalStreams = nullptr,
+                     int n_int_streams = 0, bool rowMajorIndex = true,
+                     bool rowMajorQuery = true,
+                     std::vector<int64_t> *translations = nullptr) {
+  ASSERT(DistanceType == Distance::EucUnexpandedL2 ||
+           DistanceType == Distance::EucUnexpandedL2Sqrt,
+         "Only EucUnexpandedL2Sqrt and EucUnexpandedL2 metrics are supported "
+         "currently.");
+
+  ASSERT(input.size() == sizes.size(),
+         "input and sizes vectors should be the same size");
+
+  int n_params = input.size();
+
+  std::vector<int64_t> *id_ranges;
+  if (translations == nullptr) {
+    // If we don't have explicit translations
+    // for offsets of the indices, build them
+    // from the local partitions
+    id_ranges = new std::vector<int64_t>();
+    int64_t total_n = 0;
+    for (int i = 0; i < n_params; i++) {
+      if (i < input.size()) {
+        id_ranges->push_back(total_n);
+      }
+      total_n += sizes[i];
+    }
+  } else {
+    // otherwise, use the given translations
+    id_ranges = translations;
+  }
 
-  IntType total_n = 0;
+  int device;
+  CUDA_CHECK(cudaGetDevice(&device));
 
-  for (int i = 0; i < n_params; i++) {
-    if (i < n_params)  // if i < sizes[i]
-      id_ranges->push_back(total_n);
-    total_n += sizes[i];
-  }
+  device_buffer<int64_t> trans(allocator, userStream, id_ranges->size());
+  updateDevice(trans.data(), id_ranges->data(), id_ranges->size(), userStream);
 
-  float *result_D = new float[k * size_t(n)];
-  long *result_I = new long[k * size_t(n)];
+  device_buffer<float> all_D(allocator, userStream, input.size() * k * n);
+  device_buffer<int64_t> all_I(allocator, userStream, input.size() * k * n);
 
-  float *all_D = new float[n_params * k * size_t(n)];
-  long *all_I = new long[n_params * k * size_t(n)];
+  if (n_int_streams > 0) CUDA_CHECK(cudaStreamSynchronize(userStream));
 
-  ASSERT_DEVICE_MEM(search_items, "search items");
-  ASSERT_DEVICE_MEM(res_I, "output index array");
-  ASSERT_DEVICE_MEM(res_D, "output distance array");
+  for (int i = 0; i < input.size(); i++) {
+    const float *ptr = input[i];
+    IntType size = sizes[i];
 
-  CUDA_CHECK(cudaStreamSynchronize(s));
+    try {
+      faiss::gpu::StandardGpuResources gpu_res;
 
-#pragma omp parallel
-  {
-#pragma omp for
-    for (int i = 0; i < n_params; i++) {
-      const float *ptr = input[i];
-      IntType size = sizes[i];
+      cudaStream_t stream =
+        n_int_streams > 0 ? internalStreams[i % n_int_streams] : userStream;
```

Comment:
this can be replaced with `select_stream` method below, right?

Reply:
I think so. How do you feel about including this function in the handle as well? That way we wouldn't need to explicitly pass in `n_int_streams` or `userStream`. Another option could be to have the `getInternalStream(int i)` in `cumlHandle` automatically apply a mod when `i > n_internal_streams`

Reply:
Nevermind- we don't have access to the handle in this layer.